plugins/net_plugin/net_plugin.cpp

```diff
@@ -1789,6 +1831,51 @@ namespace eosio {
       }
    }
 
+   connection_ptr sync_manager::find_next_sync_node() {
+      fc_dlog(logger, "Number connections ${s}, sync_next_expected_num: ${e}, sync_known_lib_num: ${l}",
+              ("s", my_impl->connections.number_connections())("e", sync_next_expected_num)("l", sync_known_lib_num));
+      std::deque<connection_ptr> conns;
+      my_impl->connections.for_each_block_connection([&](const auto& c) {
+         if (c->should_sync_from(sync_next_expected_num, sync_known_lib_num)) {
+            conns.push_back(c);
+         }
+      });
+      if (conns.size() > sync_peer_limit) {
+         std::partial_sort(conns.begin(), conns.begin() + sync_peer_limit, conns.end(), [](const connection_ptr& lhs, const connection_ptr& rhs) {
+            return lhs->get_net_latency_ns() < rhs->get_net_latency_ns();
+         });
+      }
+
+      fc_dlog(logger, "Valid sync peers ${s}, sync_ordinal ${so}", ("s", conns.size())("so", sync_ordinal.load()));
+
+      if (conns.empty()) {
+         return {};
+      }
+      if (conns.size() == 1) { // only one available
+         ++sync_ordinal;
+         conns.front()->sync_ordinal = sync_ordinal.load();
+         return conns.front();
+      }
+
+      ++sync_ordinal;
+      // example: sync_ordinal is 6 after inc above then there may be connections with 3,4,5.
+      // Choose from the lowest sync_ordinal of the sync_peer_limit of lowest latency, note 0 means not synced from yet
+      size_t the_one = 0;
+      uint32_t lowest_ordinal = std::numeric_limits<uint32_t>::max();
+      for (size_t i = 0; i < conns.size() && i < sync_peer_limit && lowest_ordinal != 0; ++i) {
+         uint32_t sync_ord = conns[i]->sync_ordinal;
+         fc_dlog(logger, "compare sync ords, conn: ${lcid}, ord: ${l} < ${r}, latency: ${lat}us",
+                 ("lcid", conns[i]->connection_id)("l", sync_ord)("r", lowest_ordinal)("lat", conns[i]->get_net_latency_ns()/1000));
+         if (sync_ord < lowest_ordinal) {
+            the_one = i;
+            lowest_ordinal = sync_ord;
+         }
+      }
```

Comment:
This should only be called by one connection when it finishes syncing. Although on start of sync it could certainly be called by multiple threads. There could certainly be some race conditions on actual sync ordinal, but I don't think it should cause any harm. Really doesn't matter too much on where you start syncing from as long as it is a low latency peer that has what you need.